Handle exceptions in created tasks

torarvid opened the pull request:
When create_task is called, the done_callback should check whether or not an exception was raised, and if so, re-raise it.

torarvid commented:
Hi @mosquito 👋🏻 I'm not 100% certain this change is the right fix, but I wanted to try and of course you can judge whether or not you think it makes sense.

In an application at my job that uses aio-pika, we've seen quite a few instances of Task exceptions not being handled. Typically, they look like `Task exception was never retrieved
future: <Task finished name='Task-2437' coro=<consumer() done, defined at /Users/tor/oda/trex/.venv/lib/python3.10/site-packages/aio_pika/queue.py:24> exception=[...blablabla...]`.

Adding this change makes my logs spit out this instead:
```
Exception in callback FutureStore.__on_task_done.<locals>.remover(<Task finishe...ven uuid."}')>) at /Users/tor/oda/trex/.venv/lib/python3.10/site-packages/aiormq/base.py:33
handle: <Handle FutureStore.__on_task_done.<locals>.remover(<Task finishe...ven uuid."}')>) at /Users/tor/oda/trex/.venv/lib/python3.10/site-packages/aiormq/base.py:33>
```

One reason I'm not sure if this is a 100% good fix, is that now I get two exception log outputs instead of one, with nearly identical stack traces. They start like this:
```
Traceback (most recent call last):
  File "/Users/tor/.pyenv/versions/3.10.1/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/Users/tor/oda/trex/.venv/lib/python3.10/site-packages/aiormq/base.py", line 43, in remover
    raise exc
  File "/Users/tor/oda/trex/.venv/lib/python3.10/site-packages/aio_pika/queue.py", line 30, in consumer
    return await create_task(callback, message)
  # and more that is identical in both cases...
```

```
Traceback (most recent call last):
  File "/Users/tor/.pyenv/versions/3.10.1/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/Users/tor/oda/trex/.venv/lib/python3.10/site-packages/aiormq/base.py", line 43, in remover
    raise exc
  File "/Users/tor/.pyenv/versions/3.10.1/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/Users/tor/oda/trex/.venv/lib/python3.10/site-packages/aiormq/base.py", line 43, in remover
    raise exc
  File "/Users/tor/oda/trex/.venv/lib/python3.10/site-packages/aio_pika/queue.py", line 30, in consumer
    return await create_task(callback, message)
  # and more that is identical in both cases...
```

I don't know if you can make better sense of this than me, but at least it seems appropriate to raise task exceptions when they happen 🙂

torarvid commented:
Hello, @mosquito. Have you had any time to look at this and see whether you think it makes sense or not? 🙂

mosquito commented:
@torarvid Hi, haven't looked closely yet. All free time is busy with this PR #164. If you have any ideas how to fix this case feel free to open a PR

mosquito commented:
Oh, at first I answered, then I realized that this is PR.

mosquito commented:
<summary>This patch should fix your issue I guess
<details>

```diff
diff --git a/aiormq/abc.py b/aiormq/abc.py
index 751385f..7c192ae 100644
--- a/aiormq/abc.py
+++ b/aiormq/abc.py
@@ -27,24 +27,24 @@ ExceptionType = Union[BaseException, Type[BaseException]]

 # noinspection PyShadowingNames
 class TaskWrapper:
-    __slots__ = "exception", "task"
+    __slots__ = "_exception", "task"

-    exception: Union[BaseException, Type[BaseException]]
+    _exception: Union[BaseException, Type[BaseException]]
     task: asyncio.Task

     def __init__(self, task: asyncio.Task):
         self.task = task
-        self.exception = asyncio.CancelledError
+        self._exception = asyncio.CancelledError

     def throw(self, exception: ExceptionType) -> None:
-        self.exception = exception
+        self._exception = exception
         self.task.cancel()

     async def __inner(self) -> Any:
         try:
             return await self.task
         except asyncio.CancelledError as e:
-            raise self.exception from e
+            raise self._exception from e

     def __await__(self, *args: Any, **kwargs: Any) -> Any:
         return self.__inner().__await__()
```

</details>

torarvid commented:
> This patch should fix your issue I guess
> 
> [patch omitted]

Good morning, @mosquito. Thanks for the reply, but I don't think I understand your suggestion... To my eyes, it seems it just renames `exception` to `_exception`, but perhaps there is some underlying magic that I don't understand (if for example `_exception` is somehow a "special name").

My premise for this PR is that it's possible for a task (child) to raise an exception, and for that exception to **not** be propagated to the awaiting task (parent), because the child task is not raised in the done_callback.

mosquito commented:
The main idea is the `TaskWrapper.__getattr__` will start to return task exception method instead of the it's own attribute.

torarvid commented:
@mosquito Ah, so you mean with that patch, I could simply add

```python
exc = future.exception()
if exc is not None:
    raise exc

```

and skip the `if isinstance(future, TaskWrapper)` stuff?

mosquito commented:
please rebase this against master branch

torarvid commented:
@mosquito rebased now ✅

decaz commented on aiormq/base.py:
```diff
@@ -35,6 +35,10 @@ def remover(*_: Any) -> None:
             if future in self.futures:
                 self.futures.remove(future)
 
+            exc = future.exception()
```
Is `asyncio.CancelledError` exception not handled here by intention?

torarvid replied:
No that was not by intention (I was just not aware this needed any special handling). I tried adding a workaround there now.

mosquito commented:
@torarvid could you please double check this on latest release?

mosquito commented:
@torarvid I'm glad to know that in your production it does no harm. However, to be honest, we have cases in our production when connections are broken, the network is generally not a very reliable thing, and I have been struggling with problems of this kind for several releases. It gets better every release, of course, and I encourage you to try updating to the latest version of the library.

torarvid commented:
@mosquito yeah, we have noticed improved resilience towards network blips with the later releases, so kudos for all the recent improvements 🎉

mosquito commented:
@torarvid I think these changes break something. Somewhere it's important to get a CanceledError and this is a reason when  tests has been broken. But what exactly I don't understand yet.

torarvid commented:
@mosquito that could be true, but remember: before I made the most recent change today, only a single commit was in the PR, and it didn't have any special handling of CancelledError, so it could possibly be something else?

torarvid commented:
@mosquito I found out some details by hacking `channel_max` to be `2` (instead of `2047`) [here](https://github.com/torarvid/aiormq/blob/fffa397a905f1d30442f6e2ad719f6bb585d8176/aiormq/connection.py#L471). Looks like there's some wrapping happening that probably gets worse and worse when channel_max is 2047? (I'm on very thin ice here, not quite sure what I'm doing 🤷🏻‍♂️). My pytest output for the first of these tests below:

<summary>
<details>

```bash
~/dev/aiormq on  check-for-task-exceptions [!] via 🐍 pyenv 3.11.1 
15:02:49 ❯ poetry run pytest -vv tests/test_connection.py -k test_no_free_channels -s -rP
======================================================================================= test session starts ========================================================================================
platform darwin -- Python 3.11.1, pytest-7.2.0, pluggy-1.0.0 -- /Users/tor/dev/aiormq/.venv/bin/python
cachedir: .pytest_cache
rootdir: /Users/tor/dev/aiormq
plugins: rst-0.0.7, cov-4.0.0, aiomisc-16.2.10, pylama-8.4.1
collected 237 items / 233 deselected / 4 selected                                                                                                                                                  

tests/test_connection.py::test_no_free_channels[amqp] Unexpected connection close from remote "amqp://guest:******@localhost:5672/", Connection.Close(reply_code=530, reply_text='NOT_ALLOWED - number of channels opened (2) has reached the negotiated channel_max (2)')
NoneType: None
Exception in callback <function FutureStore.__on_task_done.<locals>.remover at 0x1092eb4c0>
handle: <Handle FutureStore.__on_task_done.<locals>.remover>
Traceback (most recent call last):
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 624, in __reader
    await handler(frame)
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 557, in __handle_close
    raise exception
aiormq.exceptions.ConnectionNotAllowed: NOT_ALLOWED - number of channels opened (2) has reached the negotiated channel_max (2)
Exception in callback <function FutureStore.__on_task_done.<locals>.remover at 0x1092eb600>
handle: <Handle FutureStore.__on_task_done.<locals>.remover>
Traceback (most recent call last):
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "/usr/local/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/tasks.py", line 684, in _wrap_awaitable
    return (yield from awaitable.__await__())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/abc.py", line 44, in __inner
    return await self.task
           ^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 624, in __reader
    await handler(frame)
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 557, in __handle_close
    raise exception
aiormq.exceptions.ConnectionNotAllowed: NOT_ALLOWED - number of channels opened (2) has reached the negotiated channel_max (2)
Exception in callback <function FutureStore.__on_task_done.<locals>.remover at 0x1092ebd80>
handle: <Handle FutureStore.__on_task_done.<locals>.remover>
Traceback (most recent call last):
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "/usr/local/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/tasks.py", line 684, in _wrap_awaitable
    return (yield from awaitable.__await__())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/abc.py", line 44, in __inner
    return await self.task
           ^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 624, in __reader
    await handler(frame)
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 557, in __handle_close
    raise exception
aiormq.exceptions.ConnectionNotAllowed: NOT_ALLOWED - number of channels opened (2) has reached the negotiated channel_max (2)
Exception in callback <function FutureStore.__on_task_done.<locals>.remover at 0x1092e9bc0>
handle: <Handle FutureStore.__on_task_done.<locals>.remover>
Traceback (most recent call last):
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "/usr/local/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/tasks.py", line 684, in _wrap_awaitable
    return (yield from awaitable.__await__())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/abc.py", line 44, in __inner
    return await self.task
           ^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 624, in __reader
    await handler(frame)
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 557, in __handle_close
    raise exception
aiormq.exceptions.ConnectionNotAllowed: NOT_ALLOWED - number of channels opened (2) has reached the negotiated channel_max (2)
Exception in callback <function FutureStore.__on_task_done.<locals>.remover at 0x1092eb740>
handle: <Handle FutureStore.__on_task_done.<locals>.remover>
Traceback (most recent call last):
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "/usr/local/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/tasks.py", line 684, in _wrap_awaitable
    return (yield from awaitable.__await__())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/abc.py", line 44, in __inner
    return await self.task
           ^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 624, in __reader
    await handler(frame)
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 557, in __handle_close
    raise exception
aiormq.exceptions.ConnectionNotAllowed: NOT_ALLOWED - number of channels opened (2) has reached the negotiated channel_max (2)
Exception in callback <function FutureStore.__on_task_done.<locals>.remover at 0x1092ebba0>
handle: <Handle FutureStore.__on_task_done.<locals>.remover>
Traceback (most recent call last):
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "/usr/local/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/tasks.py", line 684, in _wrap_awaitable
    return (yield from awaitable.__await__())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/abc.py", line 44, in __inner
    return await self.task
           ^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 624, in __reader
    await handler(frame)
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 557, in __handle_close
    raise exception
aiormq.exceptions.ConnectionNotAllowed: NOT_ALLOWED - number of channels opened (2) has reached the negotiated channel_max (2)
Exception in callback <function FutureStore.__on_task_done.<locals>.remover at 0x1092eb7e0>
handle: <Handle FutureStore.__on_task_done.<locals>.remover>
Traceback (most recent call last):
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "/usr/local/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/tasks.py", line 684, in _wrap_awaitable
    return (yield from awaitable.__await__())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/abc.py", line 44, in __inner
    return await self.task
           ^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 624, in __reader
    await handler(frame)
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 557, in __handle_close
    raise exception
aiormq.exceptions.ConnectionNotAllowed: NOT_ALLOWED - number of channels opened (2) has reached the negotiated channel_max (2)
Exception in callback <function FutureStore.__on_task_done.<locals>.remover at 0x1092eb880>
handle: <Handle FutureStore.__on_task_done.<locals>.remover>
Traceback (most recent call last):
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/tor/dev/aiormq/aiormq/base.py", line 41, in remover
    raise exc
  File "/usr/local/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/tasks.py", line 684, in _wrap_awaitable
    return (yield from awaitable.__await__())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/abc.py", line 44, in __inner
    return await self.task
           ^^^^^^^^^^^^^^^
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 624, in __reader
    await handler(frame)
  File "/Users/tor/dev/aiormq/aiormq/connection.py", line 557, in __handle_close
    raise exception
aiormq.exceptions.ConnectionNotAllowed: NOT_ALLOWED - number of channels opened (2) has reached the negotiated channel_max (2)
PASSED
```
</details>

mosquito commented:
@torarvid If I correctly understood your idea, then the result should be obtained only from tasks.

torarvid commented:
@mosquito I think my idea is mostly that we should (both for Futures and Tasks) ensure that if the caller does not await the future, we don't swallow exceptions. And as far as I know, the only way to do that is to have a done_callback that checks whether or not there are any exceptions.

torarvid commented:
I did a bit more debugging.

1. I set `channel_max = 20`
2. I added a counter that prints every time we run the TaskWrapper ctor (in total, 264 tasks are created)
3. I added a counter that prints every time we run the FutureStore ctor (in total, 88 stores are created)
4. I added a (global) counter that prints every time we run the FutureStore.add method (in total, 684 calls)
5. I added a (global) counter that prints every time we run the FutureStore.__on_task_done method (in total, 684 calls)

(My command line: `poetry run pytest -vv tests/test_connection.py -k test_no_free_channels -s`)
